ambry-store/src/main/java/com.github.ambry.store/StoreManager.java

```diff
@@ -71,36 +100,73 @@ We do this by making sure that the retention period for deleted messages (which
     }
   }
 
-  public void start()
-      throws StoreException {
+  /**
+   * Start the {@link DiskManager}s for all disks on this node.
+   */
+  public void start() {
     logger.info("Starting store manager");
-    // iterate through the replicas for this node and create the stores
-    for (ReplicaId replica : replicas) {
-      // check if mount path exist
-      File file = new File(replica.getMountPath());
-      if (!file.exists()) {
-        throw new IllegalStateException("Mount path does not exist " + replica.getMountPath());
+    ScheduledExecutorService diskStartupScheduler =
+        Utils.newScheduler(Math.min(diskManagers.size(), config.storeNumDiskStartupThreads), "disk-startup-", false);
+    try {
+      Map<DiskManager, Future<Void>> futures = new HashMap<>();
+      for (final DiskManager diskManager : diskManagers) {
+        Future<Void> future = diskStartupScheduler.submit(new Callable<Void>() {
+          @Override
+          public Void call()
+              throws Exception {
+            diskManager.start();
+            return null;
+          }
+        });
+        futures.put(diskManager, future);
+      }
+      int numFailed = 0;
+      for (Map.Entry<DiskManager, Future<Void>> diskManagerAndFuture : futures.entrySet()) {
+        DiskId disk = diskManagerAndFuture.getKey().getDisk();
+        try {
+          diskManagerAndFuture.getValue().get();
+        } catch (ExecutionException e) {
+          numFailed++;
+          logger.error("Exception while starting disk manager for disk" + disk, e.getCause());
+        } catch (InterruptedException e) {
+          numFailed++;
+          logger.error("Disk manager startup for disk " + disk + " was interrupted", e);
+        }
       }
-      // Partition id is used as the id of a store.
-      String storeId = replica.getPartitionId().toString();
-      Store store =
-          new BlobStore(storeId, config, scheduler, registry, replica.getReplicaPath(), replica.getCapacityInBytes(),
-              factory, recovery, hardDelete, time);
-      store.start();
-      stores.put(replica.getPartitionId(), store);
+      if (numFailed > 0) {
+        logger.error("Could not start " + numFailed + " out of " + diskManagers.size() + " disk managers",
+            StoreErrorCodes.Initialization_Error);
+      }
+    } finally {
+      diskStartupScheduler.shutdown();
```

Comment:
awaitTermination()?

Another option to consider instead of futures and scheduler is to spin off threads for each of the disks/stores and then use `thread.join()`. Also use an `AtomicReference<Exception>` (or a list of exceptions that can be added to) to record any exceptions. Do whatever is easier and works best but the number of these threads doesn't need to be configurable - i don't see why anyone would not want max parallelism.